app/helpers/ExerciseConfig/Helper.php

```diff
@@ -185,76 +200,79 @@ public function getEnvironmentsForFiles(IExercise $exercise, array $files): arra
         $pipelines = [];
+        $pipelinesIds = [];
         foreach ($env->getPipelines() as $pipeline) {
-          $pipelineId = $pipeline->getId();
-          $pipelineEntity = $this->pipelines->findOrThrow($pipelineId);
-          $pipelineConfig = $this->loader->loadPipeline($pipelineEntity->getPipelineConfig()->getParsedPipeline());
-          $pipelines[$pipelineId] = $pipelineConfig;
+          $pipelines[] = $pipeline;
+          $pipelinesIds[] = $pipeline->getId();
         }
 
         // join pipelines and return inputs from all of them
         $inputs = []; // pairs of pipeline identifier and port indexed by variable name
         $references = []; // pairs of pipeline identifier and variable indexed by variable name
-        $this->joinPipelinesAndGetInputVariables($pipelines, $inputs, $references);
+        $this->joinPipelinesAndGetInputVariables($pipelinesIds, $inputs, $references);
 
-        // go through all inputs, references are no use to us
-        foreach ($inputs as $varName => $input) {
-          if (!$input[1]->isFile()) {
-            continue;
-          }
+        for ($i = 0; $i < count($pipelines); $i++) {
+          $pipeline = $pipelines[$i];
+          $inputPorts = $inputs[$i];
 
-          // now we have only file input ports... use them... for the glory of ReCodEx
+          // go through all inputs, references are no use to us
+          foreach ($inputPorts as $varName => $inputPort) {
+            if (!$inputPort->isFile()) {
+              continue;
+            }
 
-          // try to find variable in environment config variable table
-          $variable = $envConfig->get($varName);
-          if ($variable === null) {
-            // if variable was not found in environment config, peek into exercise config
-            $variable = $env->getPipeline($input[0])->getVariablesTable()->get($varName);
-            // TODO: get pipeline has to be removed and something else used...
-          }
+            // now we have only file input ports... use them... for the glory of ReCodEx
 
-          // somethings fishy here
-          if ($variable === null) {
-            throw new ExerciseConfigException("Variable '{$varName}' not found in environment or exercise config");
-          }
+            // try to find variable in environment config variable table
+            $variable = $envConfig->get($varName);
+            if ($variable === null) {
+              // if variable was not found in environment config, peek into exercise config
+              $variable = $pipeline->getVariablesTable()->get($varName);
+            }
 
-          // we only seek for the local file variables, not the remote ones...
-          // although port might be of type file, variables assigned to it
-          // may have remote-file type in case the file should be downloaded
-          if (!$variable->isFile()) {
-            continue;
-          }
+            // somethings fishy here
```

Comment:
It is hardly a fault of a fish that a variable has not been set. Unless you blame my code in front-end and you are mixing up fishes and beavers as both have positive inclination towards water.